nixos/modules/services/monitoring/datadog-agent.nix

```diff
@@ -0,0 +1,205 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.datadog-agent;
+
+  ddConf = {
+    dd_url              = "https://app.datadoghq.com";
+    skip_ssl_validation = "no";
+    api_key             = "";
+    collector_log_file  = "/var/log/datadog/collector.log";
+    forwarder_log_file  = "/var/log/datadog/forwarder.log";
+    dogstatsd_log_file  = "/var/log/datadog/dogstatsd.log";
+    pup_log_file        = "/var/log/datadog/pup.log";
+    confd_path          = "/etc/datadog-agent/conf.d";
+    additional_checksd  = "/etc/datadog-agent/checks.d";
+    use_dogstatsd       = "yes";
+  }
+  // optionalAttrs (cfg.logLevel != null) { log_level = cfg.logLevel; }
+  // optionalAttrs (cfg.hostname != null) { inherit (cfg) hostname; }
+  // optionalAttrs (cfg.tags != null ) { tags = concatStringsSep ", " cfg.tags; }
+  // cfg.extraConfig;
+
+  makeConfigDir = entries: mapAttrsToList (name: conf: {
+    source = pkgs.writeText (baseNameOf name) (builtins.toJSON conf);
+    target = "datadog-agent/" + name;
+  }) (filterAttrs (name: conf: conf != null) entries);
+
+  etcfiles = makeConfigDir
+    { "datadog.yaml" = ddConf;
+      "conf.d/disk.yaml" = cfg.diskConfig;
+      "conf.d/network.yaml" = cfg.networkConfig;
+      "conf.d/postgres.d/conf.yaml" = cfg.postgresqlConfig;
+      "conf.d/nginx.d/conf.yaml" = cfg.nginxConfig;
+      "conf.d/mongo.d/conf.yaml" = cfg.mongoConfig;
+      "conf.d/process.yaml" = cfg.processConfig;
+      "conf.d/jmx.yaml" = cfg.jmxConfig;
+    };
+
+in {
+  options.services.datadog-agent = {
+    enable = mkOption {
+      description = ''
+        Whether to enable the datadog-agent v6 monitoring service
+      '';
+      default = false;
+      type = types.bool;
+    };
+
+    package = mkOption {
+      default = pkgs.datadog-agent;
+      defaultText = "pkgs.datadog-agent";
+      description = ''
+        Which DataDog v6 agent package to use.
+        Override the <literal>pythonPackages</literal> argument
+        of this derivation to include more checks.
+      '';
+      type = types.package;
+    };
+
+    apiKeyFile = mkOption {
```

Comment:
API-wise we'd prefer if it was up to the user to determine whether they're fine with writing the key into a store path file or not. Moving config over from V5 is easily fixed with `builtins.toFile` but I thought I'd point it out anyways